README.md
[![Build Status](https://travis-ci.org/karupanerura/Time-Strptime.svg?branch=master)](https://travis-ci.org/karupanerura/Time-Strptime) [![Coverage Status](http://codecov.io/github/karupanerura/Time-Strptime/coverage.svg?branch=master)](https://codecov.io/github/karupanerura/Time-Strptime?branch=master)
# NAME

Time::Strptime - parse date and time string.

# SYNOPSIS

    use Time::Strptime qw/strptime/;

    # function
    my ($epoch_f, $offset_f) = strptime('%Y-%m-%d %H:%M:%S', '2014-01-01 00:00:00');

    # OO style
    my $fmt = Time::Strptime::Format->new('%Y-%m-%d %H:%M:%S');
    my ($epoch_o, $offset_o) = $fmt->parse('2014-01-01 00:00:00');

# DESCRIPTION

Time::Strptime is pure perl date and time string parser.
In other words, This is pure perl implementation a [strptime(3)](http://man.he.net/man3/strptime).

This module allows you to perform better by pre-compile the format by string.

benchmark:GMT(-0000) `dt=DateTime, ts=Time::Strptime, tp=Time::Piece`

    Benchmark: timing 100000 iterations of dt, dt(cached), tp, tp(cached), ts, ts(cached)...
            dt: 34 wallclock secs (34.23 usr +  0.02 sys = 34.25 CPU) @ 2919.71/s (n=100000)
    dt(cached): 21 wallclock secs (20.50 usr +  0.01 sys = 20.51 CPU) @ 4875.67/s (n=100000)
            tp:  1 wallclock secs ( 1.52 usr +  0.00 sys =  1.52 CPU) @ 65789.47/s (n=100000)
    tp(cached):  1 wallclock secs ( 0.61 usr +  0.00 sys =  0.61 CPU) @ 163934.43/s (n=100000)
            ts: 24 wallclock secs (24.32 usr +  0.01 sys = 24.33 CPU) @ 4110.15/s (n=100000)
    ts(cached):  1 wallclock secs ( 0.59 usr +  0.00 sys =  0.59 CPU) @ 169491.53/s (n=100000)
                   Rate       dt       ts dt(cached)        tp tp(cached) ts(cached)
    dt           2920/s       --     -29%       -40%      -96%       -98%       -98%
    ts           4110/s      41%       --       -16%      -94%       -97%       -98%
    dt(cached)   4876/s      67%      19%         --      -93%       -97%       -97%
    tp          65789/s    2153%    1501%      1249%        --       -60%       -61%
    tp(cached) 163934/s    5515%    3889%      3262%      149%         --        -3%
    ts(cached) 169492/s    5705%    4024%      3376%      158%         3%         --

benchmark:Asia/Tokyo(-0900) `dt=DateTime, ts=Time::Strptime, tp=Time::Piece`

    Benchmark: timing 100000 iterations of dt, dt(cached), tp, tp(cached), ts, ts(cached)...
            dt: 41 wallclock secs (40.74 usr +  0.02 sys = 40.76 CPU) @ 2453.39/s (n=100000)
    dt(cached): 26 wallclock secs (26.09 usr +  0.01 sys = 26.10 CPU) @ 3831.42/s (n=100000)
            tp:  2 wallclock secs ( 2.10 usr +  0.00 sys =  2.10 CPU) @ 47619.05/s (n=100000)
    tp(cached):  1 wallclock secs ( 1.48 usr +  0.01 sys =  1.49 CPU) @ 67114.09/s (n=100000)
            ts: 27 wallclock secs (26.74 usr +  0.01 sys = 26.75 CPU) @ 3738.32/s (n=100000)
    ts(cached):  1 wallclock secs ( 0.83 usr +  0.00 sys =  0.83 CPU) @ 120481.93/s (n=100000)
                   Rate       dt       ts dt(cached)        tp tp(cached) ts(cached)
    dt           2453/s       --     -34%       -36%      -95%       -96%       -98%
    ts           3738/s      52%       --        -2%      -92%       -94%       -97%
    dt(cached)   3831/s      56%       2%         --      -92%       -94%       -97%
    tp          47619/s    1841%    1174%      1143%        --       -29%       -60%
    tp(cached)  67114/s    2636%    1695%      1652%       41%         --       -44%
    ts(cached) 120482/s    4811%    3123%      3045%      153%        80%         --

# FAQ

## What's the difference between this module and other modules?

This module is fast and not require XS. but, support epoch `strptime` only.
[DateTime](https://metacpan.org/pod/DateTime) is very useful and stable! but, It is slow.
[Time::Piece](https://metacpan.org/pod/Time::Piece) is fast and useful! but, treatment of time zone is confusing. and, require XS.
[Time::Moment](https://metacpan.org/pod/Time::Moment) is very fast and useful! but, not support `strptime`. and, require XS.

## How to specify a time zone?

Set time zone name or [DateTime::TimeZone](https://metacpan.org/pod/DateTime::TimeZone) object to `time_zone` option.

    use Time::Strptime::Format;

    my $format = Time::Strptime::Format->new('%Y-%m-%d %H:%M:%S', { time_zone => 'Asia/Tokyo' });
    my ($epoch, $offset) = $format->parse('2014-01-01 00:00:00');

## How to specify a locale?

Set locale name object to `locale` option.

    use Time::Strptime::Format;

    my $format = Time::Strptime::Format->new('%Y-%m-%d %H:%M:%S', { locale => 'ja_JP' });
    my ($epoch, $offset) = $format->parse('2014-01-01 00:00:00');

# LICENSE

Copyright (C) karupanerura.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

karupanerura &lt;karupa@cpan.org>


Changes
Revision history for Perl extension Time-Strptime

{{$NEXT}}

1.03 2018-07-03T15:12:15Z

    - fixed tests on no DateTime.pm enviroments
    - fixed docs

1.02 2017-12-31T08:46:15Z

    - fixed parsing "Z" as zero offset time

1.01 2017-05-28T09:29:29Z

    - win32 uses double quote characters to quote shell args (thanks charsbar)
    - enables x_statis_install

1.00 2016-03-19T10:21:00Z

    - bump up version

1.00_01 2016-03-19T09:46:49Z

    - fixed dependency
    - updated benchmark result

0.99 2016-03-19T09:21:11Z

    - fixed something bugs
    - added documentation

0.07 2016-02-16T01:23:55Z

    - fixed dependency

0.06 2016-02-15T06:56:51Z

    - allow to omit the day of month

0.05 2016-02-15T04:21:04Z

    - allow to use for blessed DateTime::TimeZone

0.04 2016-02-15T02:03:26Z

    - time zone suppoot by DateTime::TimeZone

0.03 2015-09-04T14:58:46Z

    - Fixed utf8 flag problem on perl 5.22
    - Fixed regexp escape when charactor

0.02_01 2015-08-29T06:47:42Z

    - micro performance optimization.
    - fixed "%A", "%a" and "%C" format definition.

0.01_1 2015-01-22T10:34:02Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Time-Strptime',
    module_name     => 'Time::Strptime',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,

    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Time/Strptime.pm
package Time::Strptime;
use 5.008005;
use strict;
use warnings;

our $VERSION = "1.03";

use parent qw/Exporter/;
our @EXPORT_OK = qw/strptime/;

use Carp ();
use Time::Strptime::Format;

my %instance_cache;
sub strptime {
    my ($format_text, $date_text) = @_;

    local $Carp::CarpLevel = $Carp::CarpLevel + 1;
    my $format = $instance_cache{$format_text} ||= Time::Strptime::Format->new($format_text);
    return $format->parse($date_text);
}

1;
__END__

=encoding utf-8

=head1 NAME

Time::Strptime - parse date and time string.

=head1 SYNOPSIS

    use Time::Strptime qw/strptime/;

    # function
    my ($epoch_f, $offset_f) = strptime('%Y-%m-%d %H:%M:%S', '2014-01-01 00:00:00');

    # OO style
    my $fmt = Time::Strptime::Format->new('%Y-%m-%d %H:%M:%S');
    my ($epoch_o, $offset_o) = $fmt->parse('2014-01-01 00:00:00');

=head1 DESCRIPTION

Time::Strptime is pure perl date and time string parser.
In other words, This is pure perl implementation a L<strptime(3)>.

This module allows you to perform better by pre-compile the format by string.

benchmark:GMT(-0000) C<tp=Time::Piece, ts=Time::Strptime, pt=POSIX::strptime(+Time::Local), tm=Time::Moment>

    Benchmark: running pt, tm, tp, tp(cached), ts(cached) for at least 10 CPU seconds...
            pt: 11 wallclock secs (10.41 usr +  0.01 sys = 10.42 CPU) @ 297345.59/s (n=3098341)
            tm: 10 wallclock secs (10.17 usr +  0.01 sys = 10.18 CPU) @ 2481673.28/s (n=25263434)
            tp: 10 wallclock secs (10.52 usr +  0.01 sys = 10.53 CPU) @ 56390.98/s (n=593797)
    tp(cached): 11 wallclock secs (10.53 usr +  0.01 sys = 10.54 CPU) @ 80838.24/s (n=852035)
    ts(cached): 11 wallclock secs (10.60 usr +  0.01 sys = 10.61 CPU) @ 267686.15/s (n=2840150)
                    Rate         tp tp(cached) ts(cached)         pt         tm
    tp           56391/s         --       -30%       -79%       -81%       -98%
    tp(cached)   80838/s        43%         --       -70%       -73%       -97%
    ts(cached)  267686/s       375%       231%         --       -10%       -89%
    pt          297346/s       427%       268%        11%         --       -88%
    tm         2481673/s      4301%      2970%       827%       735%         --

benchmark:Asia/Tokyo(-0900) C<tp=Time::Piece, ts=Time::Strptime, pt=POSIX::strptime(+Time::Local), tm=Time::Moment>

    Benchmark: running pt, tm, tp, tp(cached), ts(cached) for at least 10 CPU seconds...
            pt: 10 wallclock secs (10.29 usr +  0.05 sys = 10.34 CPU) @ 147048.07/s (n=1520477)
            tm: 10 wallclock secs (10.00 usr +  0.03 sys = 10.03 CPU) @ 2344311.67/s (n=23513446)
            tp: 10 wallclock secs (10.15 usr +  0.02 sys = 10.17 CPU) @ 44565.39/s (n=453230)
    tp(cached): 11 wallclock secs (10.41 usr +  0.06 sys = 10.47 CPU) @ 50136.29/s (n=524927)
    ts(cached): 10 wallclock secs (10.73 usr +  0.07 sys = 10.80 CPU) @ 114871.48/s (n=1240612)
                    Rate         tp tp(cached) ts(cached)         pt         tm
    tp           44565/s         --       -11%       -61%       -70%       -98%
    tp(cached)   50136/s        13%         --       -56%       -66%       -98%
    ts(cached)  114871/s       158%       129%         --       -22%       -95%
    pt          147048/s       230%       193%        28%         --       -94%
    tm         2344312/s      5160%      4576%      1941%      1494%         --

=head1 FAQ

=head2 What's the difference between this module and other modules?

This module is fast and not require XS. but, support epoch C<strptime> only.
L<DateTime> is very useful and stable! but, It is slow.
L<Time::Piece> is fast and useful! but, treatment of time zone is confusing. and, require XS.
L<Time::Moment> is very fast and useful! but, does not support C<strptime>. and, require XS.

=head2 How to specify a time zone?

Set time zone name or L<DateTime::TimeZone> object to C<time_zone> option.

    use Time::Strptime::Format;

    my $format = Time::Strptime::Format->new('%Y-%m-%d %H:%M:%S', { time_zone => 'Asia/Tokyo' });
    my ($epoch, $offset) = $format->parse('2014-01-01 00:00:00');

=head2 How to specify a locale?

Set locale name object to C<locale> option.

    use Time::Strptime::Format;

    my $format = Time::Strptime::Format->new('%Y-%m-%d %H:%M:%S', { locale => 'ja_JP' });
    my ($epoch, $offset) = $format->parse('2014-01-01 00:00:00');

=head1 LICENSE

Copyright (C) karupanerura.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

karupanerura E<lt>karupa@cpan.orgE<gt>

=cut


author/benchmark.pl
use strict;
use warnings;
use utf8;
use feature qw/say/;

use Benchmark qw/cmpthese timethese/;

use Time::Strptime;
use Time::Strptime::Format;
use Time::Piece;
use DateTime::Format::Strptime;
use POSIX qw/tzset/;
use Time::Local qw/timelocal/;
use Time::TZOffset qw/tzoffset tzoffset_as_seconds/;
use Time::Moment;
use POSIX::strptime;
use Test::More;

my $pattern = '%Y-%m-%d %H:%M:%S';
my $text    = '2014-01-01 01:23:45';

say "================ Perl5 info  ==============";
system $^X, '-V';
say "================ Module info ==============";
say "$_:\t",$_->VERSION for qw/DateTime DateTime::TimeZone DateTime::Locale DateTime::Format::Strptime Time::Local Time::TZOffset Time::Moment Time::Piece Time::Strptime/;
say "===========================================";

for my $time_zone (qw|GMT UTC Asia/Tokyo America/Whitehorse|) {
    local $ENV{TZ} = $time_zone;
    tzset();

    my $ts_parser = Time::Strptime::Format->new($pattern);
    my $dt_parser = DateTime::Format::Strptime->new(pattern => $pattern, time_zone => $time_zone);
    my $tp_parser = tzoffset(CORE::localtime) eq '+0000' ? Time::Piece->gmtime : Time::Piece->localtime;

    subtest "${time_zone}(@{[ tzoffset(CORE::localtime) ]})" => sub {
        my $dt = $dt_parser->parse_datetime($text);
        my $tp = $tp_parser->strptime($text, $pattern);
        my $tm = Time::Moment->from_string($text.tzoffset(CORE::localtime), lenient => 1);
        is_deeply(($ts_parser->parse($text))[0], timelocal(POSIX::strptime($text, $pattern)));
        is_deeply([$ts_parser->parse($text)],    [$dt->epoch, $dt->offset]);
        is_deeply([$ts_parser->parse($text)],    [$tp->epoch, $tp->tzoffset->seconds]);
        is_deeply([$ts_parser->parse($text)],    [$tm->epoch, $tm->offset * 60]);
    };

    my $tzoffset = tzoffset(CORE::localtime);
    cmpthese timethese -10 => +{
        'dt(cached)' => sub { $dt_parser->parse_datetime($text) },
        'pt'         => sub { timelocal(POSIX::strptime($text, $pattern)) },
        'ts(cached)' => sub { $ts_parser->parse($text) },
        'tp(cached)' => sub { $tp_parser->strptime($text, $pattern) },
        'dt'         => sub { DateTime::Format::Strptime->new(pattern => $pattern, time_zone => $time_zone)->parse_datetime($text) },
        'ts'         => sub { Time::Strptime::Format->new($pattern)->parse($text)                                                  },
        'tp'         => sub { Time::Piece->localtime->strptime($text, $pattern) },
        'tm'         => sub { Time::Moment->from_string($text.$tzoffset, lenient => 1) },
    };
}

done_testing;
__END__
================ Perl5 info  ==============
Summary of my perl5 (revision 5 version 22 subversion 1) configuration:
   
  Platform:
    osname=darwin, osvers=14.5.0, archname=darwin-2level
    uname='darwin karupanerura-mbp.local 14.5.0 darwin kernel version 14.5.0: tue sep 1 21:23:09 pdt 2015; root:xnu-2782.50.1~1release_x86_64 x86_64 '
    config_args='-Dprefix=/Users/karupanerura/.anyenv/envs/plenv/versions/5.22 -de -Dusedevel -A'eval:scriptdir=/Users/karupanerura/.anyenv/envs/plenv/versions/5.22/bin''
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-common -DPERL_DARWIN -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include',
    optimize='-O3',
    cppflags='-fno-common -DPERL_DARWIN -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion='', gccversion='4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678, doublekind=3
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16, longdblkind=3
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/7.0.2/lib /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib /usr/lib
    libs=-lpthread -lgdbm -ldbm -ldl -lm -lutil -lc
    perllibs=-lpthread -ldl -lm -lutil -lc
    libc=, so=dylib, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' -bundle -undefined dynamic_lookup -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl): 
  Compile-time options: HAS_TIMES PERLIO_LAYERS PERL_DONT_CREATE_GVSV
                        PERL_HASH_FUNC_ONE_AT_A_TIME_HARD PERL_MALLOC_WRAP
                        PERL_NEW_COPY_ON_WRITE PERL_PRESERVE_IVUV
                        PERL_USE_DEVEL USE_64_BIT_ALL USE_64_BIT_INT
                        USE_LARGE_FILES USE_LOCALE USE_LOCALE_COLLATE
                        USE_LOCALE_CTYPE USE_LOCALE_NUMERIC USE_LOCALE_TIME
                        USE_PERLIO USE_PERL_ATOF
  Locally applied patches:
	Devel::PatchPerl 1.38
  Built under darwin
  Compiled at Dec 14 2015 13:05:08
  @INC:
    /Users/karupanerura/.anyenv/envs/plenv/versions/5.22/lib/perl5/site_perl/5.22.1/darwin-2level
    /Users/karupanerura/.anyenv/envs/plenv/versions/5.22/lib/perl5/site_perl/5.22.1
    /Users/karupanerura/.anyenv/envs/plenv/versions/5.22/lib/perl5/5.22.1/darwin-2level
    /Users/karupanerura/.anyenv/envs/plenv/versions/5.22/lib/perl5/5.22.1
    .
================ Module info ==============
DateTime:	1.21
DateTime::TimeZone:	1.94
DateTime::Locale:	1.02
DateTime::Format::Strptime:	1.62
Time::Local:	1.2300
Time::TZOffset:	0.04
Time::Moment:	0.37
Time::Piece:	1.29
Time::Strptime:	0.03
===========================================
    # Subtest: GMT(+0000)
    ok 1
    ok 2
    ok 3
    ok 4
    1..4
ok 1 - GMT(+0000)
Benchmark: running dt, dt(cached), pt, tm, tp, tp(cached), ts, ts(cached) for at least 10 CPU seconds...
        dt: 11 wallclock secs (10.45 usr +  0.01 sys = 10.46 CPU) @ 2313.48/s (n=24199)
dt(cached): 11 wallclock secs (10.55 usr +  0.01 sys = 10.56 CPU) @ 7906.72/s (n=83495)
        pt: 11 wallclock secs (10.48 usr +  0.00 sys = 10.48 CPU) @ 186033.21/s (n=1949628)
        tm: 11 wallclock secs (10.40 usr +  0.00 sys = 10.40 CPU) @ 1927168.46/s (n=20042552)
        tp: 11 wallclock secs (10.53 usr +  0.00 sys = 10.53 CPU) @ 116695.06/s (n=1228799)
tp(cached): 11 wallclock secs (10.60 usr +  0.01 sys = 10.61 CPU) @ 286613.10/s (n=3040965)
        ts: 10 wallclock secs (10.24 usr +  0.35 sys = 10.59 CPU) @ 2846.74/s (n=30147)
ts(cached): 11 wallclock secs (10.37 usr +  0.00 sys = 10.37 CPU) @ 172462.97/s (n=1788441)
                Rate     dt     ts dt(cached)    tp ts(cached)   pt tp(cached)    tm
dt            2313/s     --   -19%       -71%  -98%       -99% -99%       -99% -100%
ts            2847/s    23%     --       -64%  -98%       -98% -98%       -99% -100%
dt(cached)    7907/s   242%   178%         --  -93%       -95% -96%       -97% -100%
tp          116695/s  4944%  3999%      1376%    --       -32% -37%       -59%  -94%
ts(cached)  172463/s  7355%  5958%      2081%   48%         --  -7%       -40%  -91%
pt          186033/s  7941%  6435%      2253%   59%         8%   --       -35%  -90%
tp(cached)  286613/s 12289%  9968%      3525%  146%        66%  54%         --  -85%
tm         1927168/s 83202% 67597%     24274% 1551%      1017% 936%       572%    --
    # Subtest: UTC(+0000)
    ok 1
    ok 2
    ok 3
    ok 4
    1..4
ok 2 - UTC(+0000)
Benchmark: running dt, dt(cached), pt, tm, tp, tp(cached), ts, ts(cached) for at least 10 CPU seconds...
        dt: 10 wallclock secs (10.66 usr +  0.01 sys = 10.67 CPU) @ 2267.95/s (n=24199)
dt(cached): 10 wallclock secs (10.54 usr +  0.01 sys = 10.55 CPU) @ 7692.32/s (n=81154)
        pt: 11 wallclock secs (10.40 usr +  0.01 sys = 10.41 CPU) @ 185383.19/s (n=1929839)
        tm: 10 wallclock secs (10.48 usr +  0.00 sys = 10.48 CPU) @ 1928506.01/s (n=20210743)
        tp: 10 wallclock secs (10.03 usr +  0.00 sys = 10.03 CPU) @ 130361.62/s (n=1307527)
tp(cached): 11 wallclock secs (10.59 usr +  0.01 sys = 10.60 CPU) @ 286883.49/s (n=3040965)
        ts: 10 wallclock secs ( 9.72 usr +  0.34 sys = 10.06 CPU) @ 2831.01/s (n=28480)
ts(cached): 10 wallclock secs (10.33 usr +  0.00 sys = 10.33 CPU) @ 171433.40/s (n=1770907)
                Rate     dt     ts dt(cached)    tp ts(cached)   pt tp(cached)    tm
dt            2268/s     --   -20%       -71%  -98%       -99% -99%       -99% -100%
ts            2831/s    25%     --       -63%  -98%       -98% -98%       -99% -100%
dt(cached)    7692/s   239%   172%         --  -94%       -96% -96%       -97% -100%
tp          130362/s  5648%  4505%      1595%    --       -24% -30%       -55%  -93%
ts(cached)  171433/s  7459%  5956%      2129%   32%         --  -8%       -40%  -91%
pt          185383/s  8074%  6448%      2310%   42%         8%   --       -35%  -90%
tp(cached)  286883/s 12549% 10034%      3629%  120%        67%  55%         --  -85%
tm         1928506/s 84933% 68021%     24971% 1379%      1025% 940%       572%    --
    # Subtest: Asia/Tokyo(+0900)
    ok 1
    ok 2
    ok 3
    ok 4
    1..4
ok 3 - Asia/Tokyo(+0900)
Benchmark: running dt, dt(cached), pt, tm, tp, tp(cached), ts, ts(cached) for at least 10 CPU seconds...
        dt: 10 wallclock secs (10.56 usr +  0.01 sys = 10.57 CPU) @ 2159.32/s (n=22824)
dt(cached): 11 wallclock secs (10.55 usr +  0.01 sys = 10.56 CPU) @ 6756.91/s (n=71353)
        pt: 11 wallclock secs (10.51 usr +  0.00 sys = 10.51 CPU) @ 113668.51/s (n=1194656)
        tm: 12 wallclock secs (10.58 usr + -0.01 sys = 10.57 CPU) @ 1933722.33/s (n=20439445)
        tp: 10 wallclock secs (10.30 usr +  0.01 sys = 10.31 CPU) @ 116956.45/s (n=1205821)
tp(cached):  9 wallclock secs (10.48 usr +  0.01 sys = 10.49 CPU) @ 287232.22/s (n=3013066)
        ts: 11 wallclock secs (10.19 usr +  0.33 sys = 10.52 CPU) @ 2614.45/s (n=27504)
ts(cached):  9 wallclock secs (10.37 usr +  0.01 sys = 10.38 CPU) @ 84805.30/s (n=880279)
                Rate     dt     ts dt(cached) ts(cached)    pt    tp tp(cached)    tm
dt            2159/s     --   -17%       -68%       -97%  -98%  -98%       -99% -100%
ts            2614/s    21%     --       -61%       -97%  -98%  -98%       -99% -100%
dt(cached)    6757/s   213%   158%         --       -92%  -94%  -94%       -98% -100%
ts(cached)   84805/s  3827%  3144%      1155%         --  -25%  -27%       -70%  -96%
pt          113669/s  5164%  4248%      1582%        34%    --   -3%       -60%  -94%
tp          116956/s  5316%  4373%      1631%        38%    3%    --       -59%  -94%
tp(cached)  287232/s 13202% 10886%      4151%       239%  153%  146%         --  -85%
tm         1933722/s 89452% 73863%     28518%      2180% 1601% 1553%       573%    --
    # Subtest: America/Whitehorse(-0800)
    ok 1
    ok 2
    ok 3
    ok 4
    1..4
ok 4 - America/Whitehorse(-0800)
Benchmark: running dt, dt(cached), pt, tm, tp, tp(cached), ts, ts(cached) for at least 10 CPU seconds...
        dt: 10 wallclock secs (10.56 usr +  0.01 sys = 10.57 CPU) @ 2147.59/s (n=22700)
dt(cached): 10 wallclock secs (10.55 usr +  0.00 sys = 10.55 CPU) @ 6794.22/s (n=71679)
        pt: 11 wallclock secs (10.46 usr +  0.00 sys = 10.46 CPU) @ 94936.42/s (n=993035)
        tm: 11 wallclock secs (10.65 usr +  0.01 sys = 10.66 CPU) @ 1917396.34/s (n=20439445)
        tp: 10 wallclock secs (10.50 usr +  0.01 sys = 10.51 CPU) @ 116441.10/s (n=1223796)
tp(cached):  9 wallclock secs (10.42 usr +  0.00 sys = 10.42 CPU) @ 286533.01/s (n=2985674)
        ts: 11 wallclock secs (10.17 usr +  0.33 sys = 10.50 CPU) @ 2609.62/s (n=27401)
ts(cached):  9 wallclock secs (10.56 usr +  0.01 sys = 10.57 CPU) @ 87907.38/s (n=929181)
                Rate     dt     ts dt(cached) ts(cached)    pt    tp tp(cached)    tm
dt            2148/s     --   -18%       -68%       -98%  -98%  -98%       -99% -100%
ts            2610/s    22%     --       -62%       -97%  -97%  -98%       -99% -100%
dt(cached)    6794/s   216%   160%         --       -92%  -93%  -94%       -98% -100%
ts(cached)   87907/s  3993%  3269%      1194%         --   -7%  -25%       -69%  -95%
pt           94936/s  4321%  3538%      1297%         8%    --  -18%       -67%  -95%
tp          116441/s  5322%  4362%      1614%        32%   23%    --       -59%  -94%
tp(cached)  286533/s 13242% 10880%      4117%       226%  202%  146%         --  -85%
tm         1917396/s 89181% 73374%     28121%      2081% 1920% 1547%       569%    --
1..4
